template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -186,6 +187,20 @@ public void addVariables(List<ApiClusterTemplateVariable> vars) {
         }
     }
 
+    public void extendTemplateWithAdditionalServices(Map<String, ApiClusterTemplateService> hostGroupServices) {
+        for (Entry<String, ApiClusterTemplateService> hostGroupService : hostGroupServices.entrySet()) {
+            ApiClusterTemplateService service = hostGroupService.getValue();
+            List<String> serviceRefNames = service.getRoleConfigGroups().stream()
+                    .map(ApiClusterTemplateRoleConfigGroup::getRefName).collect(Collectors.toList());
+            if (getServiceByType(service.getServiceType()).isEmpty()) {
+                cmTemplate.addServicesItem(service);
```

Comment:
The content of the service (configs, role config groups) may need to be merged with the existing service of the same type.

Reply:
Yeah, the number of combinations that a blueprint can have is so big, I don't want to cover all the scenarios. Otherwise I would need to consider:
* service does not exist at all
* service exists, but not the specific role
* service exists, with 1 or more non base roles
* service exists, but no base role exists

Right now if the service exists it is not going to add it (even if the specific role is missing). So far only Knox had to be added to the blueprint and there were no such requirement for any other services. That's why it is tailored to Knox right now and not for other services.